### PR TITLE
#1965: improve dependent installation

### LINKED ISSUE #1965: Improve dependent installations
### Task summary

We already support tool dependencies via `dependencies.json` files from `ide-urls`.
This is properly handled here:
https://github.com/devonfw/IDEasy/blob/5cf051aece16d55a76211ee2b7b1465c5af6cdee/cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java#L276-L301

As we can see we are creating a `ToolInstallationRequest` for the dependent installation as a child that is connected to the parent `ToolInstallationRequest`.
This way we can properly detect cycles in dependent installations that could lead to infinity loops:
https://github.com/devonfw/IDEasy/blob/5cf051aece16d55a76211ee2b7b1465c5af6cdee/cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java#L158-L163

However, we also have `PackageManagerBasedLocalToolCommandlet` that implicitly causes a dependency to the package-manager (e.g. `npm` or `pip`).
Here we had infinity loop bugs causing stack-overflow errors in tests.
As a hackish quickfix we have this:
https://github.com/devonfw/IDEasy/blob/5cf051aece16d55a76211ee2b7b1465c5af6cdee/cli/src/main/java/com/devonfw/tools/ide/tool/PackageManagerBasedLocalToolCommandlet.java#L88-L99

and:
https://github.com/devonfw/IDEasy/blob/5cf051aece16d55a76211ee2b7b1465c5af6cdee/cli/src/main/java/com/devonfw/tools/ide/tool/node/Node.java#L47

The idea of this story is to improve this and create a clean solution replacing the hack.
The problem is actually already visible to end-users via our logs:
```
$ ide install nest
No CVEs found for version 11.15.0 of tool npm.
No CVEs found for version v24.15.0 of tool node.
...
No CVEs found for version 11.0.21 of tool nest.
No CVEs found for version 11.15.0 of tool npm.
No CVEs found for version v24.15.0 of tool node.
...
Successfully installed nest in version 11.0.21 replacing previous version v24.15.0 of nest at D:\projects\test\software\node
```
So still we see that installation of `npm` and `node` was called twice because loop was not detected.
After this issue was fixed we should only see:
```
$ ide install nest
No CVEs found for version 11.15.0 of tool npm.
No CVEs found for version v24.15.0 of tool node.
...
No CVEs found for version 11.0.21 of tool nest.
...
Successfully installed nest in version v24.15.0 of nest at D:\projects\test\software\node
```

### Additional context

In order to implement a clean fix we would somehow refactor `PackageManagerRequest` to somehow be compatbile with `ToolInstallRequest` or to have the ability to contain a `ToolInstallRequest` as a parent.
Then we could use the same chaining and loop detection fixing this issue.

### PR DESCRIPTION
**Pull request by MeShehi**
### Implemented changes:

* Fixed npm and pip-based installation to avoid double installation cycle
* Added ToolInstallRequest as a parent of PackageManagerRequest, added possibility to use PackageManagerRequest as a parent to ToolinstallRequest
---

### Testing instructions

1. Create or edit your configuration to run `-cp ide-cli` with `ide uninstall npm` and run, wait to end
2. Run the same configuration with `ide install nest`
3. No duplication should be present by the logs shown in terminal

---

### Checklist for this PR

Make sure everything is checked before merging this PR. For further info please also see
our [DoD](https://github.com/devonfw/IDEasy/blob/main/documentation/contributing/DoD.adoc).

- [x] When running `mvn clean test` locally all tests pass and build is successful
- [x] PR title is of the form `#«issue-id»: «brief summary»` (e.g. `#921: fixed setup.bat`). If no issue ID exists, title only.
- [x] PR top-level comment summarizes what has been done and contains link to addressed issue(s)
- [x] PR and issue(s) have suitable labels
- [x] Issue is set to `In Progress` and assigned to you *or* there is no issue (might happen for very small PRs)
- [x] You followed all [coding conventions](https://github.com/devonfw/IDEasy/blob/main/documentation/coding-conventions.adoc)
- [x] You have added the issue implemented by your PR in [CHANGELOG.adoc](https://github.com/devonfw/IDEasy/blob/main/CHANGELOG.adoc) unless issue is labeled
  with `internal`
- [x] You have formulated clear instructions on how to test your contribution under "Testing instructions"

### CODE REVIEW

**Comment by MeShehi** [comment]
Please, when checking the code give extra attention to ToolInstallRequest.detectInstallLoopRecursively, since that's where I suspect the new changes might have unsuspected failures. The tests do pass, but maybe edge cases are not covered

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolInstallRequest.java [hunk 1]
```diff
@@ -106,10 +129,19 @@ private boolean detectInstallLoopRecursively(ToolEditionAndVersion toolEditionAn
...
+
+    boolean loopFound;
+
+    if (this.parentToolInstallRequest != null) {
+      loopFound = this.parentToolInstallRequest.detectInstallLoopRecursively(toolEditionAndVersion, sb);
+    } else if (this.parentPackageManagerRequest != null) {
+      if (this.parentPackageManagerRequest.getType().equals("uninstall")) {
+        return false;
+      }
+      loopFound = this.parentPackageManagerRequest.getToolInstallRequest().detectInstallLoopRecursively(toolEditionAndVersion, sb);
```
Comment (quando632) [inline cli/src/main/java/com/devonfw/tools/ide/tool/ToolInstallRequest.java]: `getToolInstallRequest()` is null for every request that carries no tool installation, so only the hard-coded `"uninstall"` comparison keeps this from dereferencing null. The public one-arg `runPackageManager` makes other types reachable. Guarding on the parent itself covers the uninstall case too, since those requests never carry one.

```suggestion
    } else if (this.parentPackageManagerRequest != null) {
      // a package manager request without tool installation (e.g. uninstall) cannot be part of an installation loop
      ToolInstallRequest parentRequest = this.parentPackageManagerRequest.getToolInstallRequest();
      if (parentRequest == null) {
        return false;
      }
      loopFound = parentRequest.detectInstallLoopRecursively(toolEditionAndVersion, sb);
```

**Review by quando632 — COMMENTED** [review]
### What works well

The root cause is correctly identified. `runPackageManager` built the nested install request without a parent, so `isInstallLoop()` could never see the cycle. A repeated `ide install nest` now skips the extra npm installation cycle that ran before, which is what the issue asks for.

`NpmBasedCommandlet.java:68` and `PipBasedCommandlet.java:113` also bring both callers in line with the contract already documented on `computeInstalledVersion`.

### Findings

**1. `PackageManagerBasedLocalToolCommandlet.java:94` no longer uses `pc.createChild()`.** The nested install now shares the caller's `ProcessContext`, including `errorHandling` and `exitCodeAcceptor`, which the child deliberately resets (`ProcessContextImpl.java:84`). Restoring it has to happen inside the constructor, since `setProcessContext` rejects a second write.

**2. `ToolInstallRequest.java:69` hard-codes `silent` to `false`, the replaced code passed `true`.** With the current `ng` fixture this moves the "already installed" lines for node and npm from DEBUG to INFO. Was that intended, or should the flag come from the parent request?

**3. The fixed duplication has no test.** `computeInstalledVersion` is only reached on a repeated installation, so a second `install()` in `NgTest` pins it. The mock npm in the `ng` fixture needs a `list` branch for this, it currently answers `--version` only:

```java
commandlet.install();
context.getTestStartContext().getEntries().clear();
commandlet.install();
assertThat(context).log().hasNoMessageContaining("No CVEs found for version 9.9.2 of tool npm");
```

**4. Step 1 of the testing instructions has no effect.** `ide uninstall npm` refuses the uninstall because `Npm.canBeUninstalled()` is `false`, and still prints "Successfully uninstalled npm". Running `ide install nest` twice reproduces the duplication reliably and would be the clearer instruction.

### Optional: one parent field instead of two

Both inline bugs come from two mutually exclusive parent fields with two constructors that must be kept in sync. Resolving the parent in `runPackageManager` via `request.getToolInstallRequest()` keeps a single field, and the nested request then inherits `silent` and `ignoreProject` automatically.

### Nits

- `ToolInstallRequest.java:90`: the private constructor has no javadoc, and its `silent` and `direct` parameters are always `false` at the single call site.
- `PackageManagerBasedLocalToolCommandlet.java:93` lost the `// See Node.postInstallOnNewInstallation` pointer that explained the npm/node cycle.

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolInstallRequest.java [hunk 2]
```diff
@@ -75,6 +87,17 @@ private ToolInstallRequest(ToolInstallRequest parent, boolean silent, boolean di
...
 
+  private ToolInstallRequest(PackageManagerRequest parent, boolean silent, boolean direct) {
+    super();
+    this.parentPackageManagerRequest = parent;
+    this.parentToolInstallRequest = null;
+    this.silent = silent;
+    this.direct = direct;
+    if (parent != null) {
+      this.processContext = parent.getProcessContext();
+    }
```
Comment (quando632) [inline cli/src/main/java/com/devonfw/tools/ide/tool/ToolInstallRequest.java:102]: This constructor does not propagate `ignoreProject`, while the constructor above it does (line 86). An install request that ignores the project therefore loses that flag as soon as it goes through a package manager, silently and without any error.

```suggestion
    if (parent != null) {
      this.processContext = parent.getProcessContext();
      ToolInstallRequest parentRequest = parent.getToolInstallRequest();
      if (parentRequest != null) {
        this.ignoreProject = parentRequest.ignoreProject;
      }
    }
```

File: CHANGELOG.adoc [hunk 3]
```diff
@@ -6,6 +6,8 @@ This file documents all notable changes to https://github.com/devonfw/IDEasy[IDE
 
 Release with new features and bugfixes:
 
+* https://github.com/devonfw/IDEasy/issues/1965[#1965]: Improve dependent installation, fixed npm and pip-based installation duplication
+* https://github.com/devonfw/IDEasy/issues/2100[#2100]: Fix Python not available for Mac x64
```
Comment (quando632) [inline CHANGELOG.adoc]: The `#2100` entry already exists further down in the same release block, added by PR #2152.

```suggestion
* https://github.com/devonfw/IDEasy/issues/1965[#1965]: Improve dependent installation, fixed npm and pip-based installation duplication
```

File: cli/src/main/java/com/devonfw/tools/ide/tool/PackageManagerRequest.java [hunk 4]
```diff
@@ -166,4 +168,26 @@ public PackageManagerRequest setProcessMode(ProcessMode processMode) {
...
+  public ToolInstallRequest getToolInstallRequest() {
+    return this.toolInstallRequest;
+  }
+
+  /**
+   * @param toolInstallRequest new value of {@link #getToolInstallRequest()} ToolInstallRequest()}.
+   * @return this {@link PackageManagerRequest} for fluent API calls.
+   */
+  public PackageManagerRequest setToolInstallRequest(
+      ToolInstallRequest toolInstallRequest) {
```
Comment (quando632) [inline cli/src/main/java/com/devonfw/tools/ide/tool/PackageManagerRequest.java]: The javadoc contains a copy/paste artifact, and the signature fits on one line within the 160 character limit.

```suggestion
  /**
   * @param toolInstallRequest new value of {@link #getToolInstallRequest()}.
   * @return this {@link PackageManagerRequest} for fluent API calls.
   */
  public PackageManagerRequest setToolInstallRequest(ToolInstallRequest toolInstallRequest) {
```

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolInstallRequest.java [hunk 5]
```diff
@@ -58,6 +60,15 @@ public ToolInstallRequest(ToolInstallRequest parent) {
     this(parent, parent.silent, false);
   }
 
+  /**
+   * The constructor.
+   *
+   * @param parent the parent {@link PackageManagerRequest} (in case of a dependency).
+   */
+  public ToolInstallRequest(PackageManagerRequest parent) {
+    this(parent, false, false);
```
Comment (quando632) [inline cli/src/main/java/com/devonfw/tools/ide/tool/ToolInstallRequest.java]: `silent` is hard-coded to `false` here, while the replaced code in `PackageManagerBasedLocalToolCommandlet` passed `true`. That moves "Version 9.9.2 of tool npm is already installed" and the node equivalent from DEBUG to INFO on every `ide install ng`, verified by running the same assertion against this branch and against the merge base.

```suggestion
    this(parent, true, false);
```

File: cli/src/main/java/com/devonfw/tools/ide/tool/npm/NpmBasedCommandlet.java [hunk 6]
```diff
@@ -65,7 +65,7 @@ private VersionIdentifier runPackageManagerGetInstalledVersion(String npmPackage
     ProcessContext pc = this.context.newProcess().errorHandling(ProcessErrorHandling.THROW_CLI)
         .withExitCodeAcceptor(rc -> true); // if the tool is not installed npm list will end with exit code 1
     request.setProcessContext(pc);
-    ProcessResult result = runPackageManager(request);
+    ProcessResult result = runPackageManager(request, true);
```
Comment (quando632) [inline cli/src/main/java/com/devonfw/tools/ide/tool/npm/NpmBasedCommandlet.java:68]: `skipInstallation=true` is only reached on a repeated installation, so nothing currently covers the fix. A second `install()` in `NgTest` pins it, but the mock npm in `ide-projects/ng/repository/node/node/default/npm` needs a `list` branch first: its `echo "npm $*"` fallback is parsed as a version and `VersionIdentifier.of` throws. With both in place the test fails on the merge base and passes here.

```bash
if [ "$1" == "list" ]; then
  echo "/lib"
  echo "|-- $3@18.0.1"
  exit
fi
```

```java
commandlet.install();
context.getTestStartContext().getEntries().clear();
commandlet.install();
assertThat(context).log().hasNoMessageContaining("No CVEs found for version 9.9.2 of tool npm");
```

**Comment by quando632** [comment]
LGTM now you only need to resolve the conflicts that it should be ready to merge

File: cli/src/main/java/com/devonfw/tools/ide/tool/ToolInstallRequest.java [hunk 7]
```diff
@@ -18,7 +18,9 @@ public final class ToolInstallRequest {
 
   private static final Logger LOG = LoggerFactory.getLogger(ToolInstallRequest.class);
 
-  private final ToolInstallRequest parent;
+  private final PackageManagerRequest parentPackageManagerRequest;
+
+  private final ToolInstallRequest parentToolInstallRequest;
```
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/tool/ToolInstallRequest.java]: This looks a little odd to me.
> we would somehow refactor PackageManagerRequest to somehow be compatbile with ToolInstallRequest or to have the ability to contain a ToolInstallRequest as a parent.

I was maybe not clear enough but I would have created a common interface or super-class or my other suggestion was to always create a "virtual instance" of `ToolInstallRequest` for a `PackageManagerRequest` representing an installation. Then we still have only one `parent` making the code easier to read, understand and to avoid bugs. WDYT?

**Review by hohwille — APPROVED** [review]
@MeShehi thanks for your PR. 👍 
I would like to ask you to improve the design.
It is fine that `PackageManagerRequest` contains a `ToolInstallRequest`.
Now if every `PackageManagerRequest` that triggers a tool installation (e.g. for `ide install ng` we trigger `npm install -gf @angular/cli@22.2.0`) already contains a `ToolInstallRequest` - that one should point to the actual installation of the new tool to install (in that example `ng`). Now if we already have a parent tool installation causing this, then this one should be the parent.
Hence, you do not need to modify `ToolInstallRequest` and it does not have to know about `PackageManagerRequest` what otherwise also causes a cyclic dependency and if that can be avoided this is always preferred for good design.
